### PR TITLE
Issue #3117677 by makishima: Add Follow taxonomy functionality and Follow tag module

### PR DESCRIPTION
**Pull request by VolodymyrMartsynenko**
## Problem
Should be possible to follow/unfollow taxonomy on wherever it appears.

## Solution
Implementing Social Follow Taxonomy module with the basic structure and functionality. Also implement Social Follow Tag and Social Follow Landing Page modules to extend basic functionalities on Tags and Landing pages (paragraphs).

## Issue tracker
Drupal.org issue: https://www.drupal.org/project/social/issues/3117677
Task #1: https://getopensocial.atlassian.net/browse/YANG-2047
Task #2: https://getopensocial.atlassian.net/browse/YANG-3464

##  HTT
- [ ] As a CM+ create new Event/Discussion/Topic (or edit an existing).
- [ ] Fill tags field and save.

- [ ] **Check functionality on the Event/Discussion/Topic pages:**

1. Login as LU and go to node you previously created/edited.
2. You should see Tags block with selected tags.
3. Click on any tag - the popup should appear.
3. In the popup you should see tag name, list of related content with counters, “Follow” button, “See related content” link.
4. Click on the Follow button, wait for ajax request, button should be replaced with Followed and the star icon should be added to tag under popup.
5. Click on the Followed button - you should see Unfollow button in dropdown.
6. Click on the Unfollow button, wait for ajax request, button should be replaced with Follow button and the star icon should be removed from tag under popup.

- [ ] **Check functionality on the Search content page:**

1. Click on See related content link (appeared on the popup) - you should be redirected to Search content page with tag selected in filters.
2. You should see block with the tag you followed on the page.
3. You should see Followed button and should be able to unfollow the tag.

- [ ] **Check functionality on the Followed tags page:**

1. Click on the user profile icon - in user menu you should see Following tags menu item, click on it and you should be redirected to the /following-tags page.
2. There you should see list of tags which you already followed.
3. You should be able to unfollow tags by pressing Followed -> Unfollow button in action column.

- [ ] **Check functionality on the Landing page**

1. As a CM+ create Landing page (or edit an existing).
2. Add a new section: “Add Tag”.
3. Select some tags from the Tag field and save the landing page.
4. On the created page you should see Tag block with the tags you selected.
5. On the Tag block you should be able to follow/unfollow tags go to related content by clicking on the “See related content” link.
‌

## Screenshots
tag_follow_02
tag_follow_04
landing and dashboard pages

### CODE REVIEW

File: modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.module [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+/**
+ * Extends variables for social_tagging_split template.
+ *
+ * Implements hook_preprocess_HOOK().
+ * {@inheritdoc}
+ */
+function social_follow_tag_preprocess_social_tagging_split(&$variables) {
+  // Iterate over an array with parent terms (category).
+  foreach ($variables['taghierarchy'] as $parent_tag) {
```
Comment (ribel) [inline modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.module]: @VolodymyrMartsynenko  Are we sure that an array always will be passed to `$variables['taghierarchy']`?
If not - maybe good to add some additional check here.

File: modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.module [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+/**
+ * Extends variables for social_tagging_nosplit template.
+ *
+ * Implements hook_preprocess_HOOK().
+ * {@inheritdoc}
+ */
+function social_follow_tag_preprocess_social_tagging_nosplit(&$variables) {
+  $term_storage = \Drupal::entityTypeManager()->getStorage('taxonomy_term');
+
+  foreach ($variables['tags'] as $tag_name => $tag_info) {
```
Comment (ribel) [inline modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.module]: @VolodymyrMartsynenko  Are we sure that an array always will be passed to `$variables['tags']`?
If not - maybe good to add some additional check here.

File: modules/social_features/social_follow_taxonomy/social_follow_taxonomy.module [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+ * Implements hook_preprocess_activity().
+ * {@inheritdoc}
+ */
+function social_follow_taxonomy_preprocess_activity(&$variables) {
+  $activity = $variables['elements']['#activity'];
+
+  $message = Message::load($activity->field_activity_message->target_id);
+  $message_template_id = $message->getTemplate()->id();
+
+  if ($message_template_id == 'create_node_following_tag') {
```
Comment (ribel) [inline modules/social_features/social_follow_taxonomy/social_follow_taxonomy.module]: @VolodymyrMartsynenko Could you give more comments about what happens here?

**Comment by ribel** [comment]
@VolodymyrMartsynenko please fix coding standards by removing unused use statements in `social_follow_tag.module`

File: modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.module (hunk 2 above)
Comment (VolodymyrMartsynenko) [reply modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.module]: I added an additional check for this.

File: modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.module (hunk 1 above)
Comment (VolodymyrMartsynenko) [reply modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.module]: I added an additional check for this.

File: modules/social_features/social_follow_taxonomy/social_follow_taxonomy.module (hunk 3 above)
Comment (VolodymyrMartsynenko) [reply modules/social_features/social_follow_taxonomy/social_follow_taxonomy.module]: I added an additional comments for this.

**Comment by ronaldtebrake** [comment]
Tests are failing, probably just needs to have base branch merged back in here

**Comment by ronaldtebrake** [comment]
Great job everybody!! This will be such an improvement and a lot of flexibility.
Some remarks from my side.

1. If I just enable social_follow_taxonomy it doesn't really do anything. What is the benefit for users to have this enabled without the social_follow_tag?
2. We can't install it, it has a dependency on Paragraphs but that isn't declared in the info.yml or anywhere. Probably this is related to needing social_landing_page, but in that case we need to put it in the optional module with correct dependencies.
 
`Drupal\Core\Config\UnmetDependenciesException: Configuration objects provided by <em class="placeholder">social_follow_tag</em> have unmet dependencies: <em class="placeholder">core.entity_form_display.block_content.follow_tags.default (paragraphs),                  [error]
core.entity_view_display.block_content.follow_tags.default (entity_reference_revisions), field.storage.block_content.field_tags (entity_reference_revisions, paragraphs)</em> in /var/www/html/core/lib/Drupal/Core/Config/UnmetDependenciesException.php:98
Stack trace:
`

3.  If you update the URL after searching with a non existent term id f.e. http://192.168.99.100:32775/search/content?parent_term%5B%5D=10 it will crash.

4. The search result doesn't always show up, while it does work when going to the search itself and it does show there should be 1 item. 
Screenshot 2020-07-29 at 21 19 07

Screenshot 2020-07-29 at 21 17 38

Screenshot 2020-07-29 at 21 20 38

5. We might want to improve this UX a bit for a SM/CM. - NONE - doesn't say that much and is added by default.
Screenshot 2020-07-29 at 21 23 39

6. From the stream, unfollowing doesn't seem to work really well. Pressing it, refreshing still has me following the tag. On the search page / the node tag block list the unfollowing does work well.
Screenshot 2020-07-29 at 21 36 48

7. We have a little information disclosure issue. 
Screenshot 2020-07-29 at 21 35 00
It shows 4 items here, while 2 of them are in a closed group where the current LU is not a member of. 
This probably also means that only 2 content items are shown. However there are some issues with that see point 4.

8. Notifications didnt seem to get triggered. Not sure when or what should be triggerd, couldnt find this information.

All in all this feature will really be a big part of the community, I think it's getting very close to being finished, well done!!

File: modules/social_features/social_follow_taxonomy/modules/social_follow_landing_page/social_follow_landing_page.info.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+name: 'Social Follow Landing Page'
+description: 'Provide follow functionality for landing pages.'
+type: module
+core: 8.x
+dependencies:
+  - social_follow_taxonomy:social_follow_taxonomy
+  - social_tagging:social_tagging
+  - social_landing_page:social_landing_page
+package: Social
```
Comment (ribel) [inline modules/social_features/social_follow_taxonomy/modules/social_follow_landing_page/social_follow_landing_page.info.yml:9]: @denysyukm I think it's better to add it to Social (experimental) package

File: modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.info.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+name: 'Social Follow Tag'
+description: 'Provide follow functionality for social_tagging module.'
+type: module
+core: 8.x
+dependencies:
+  - social_follow_taxonomy:social_follow_taxonomy
+  - social_tagging:social_tagging
+package: Social
```
Comment (ribel) [inline modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.info.yml:8]: @denysyukm I think we need to add it to the experimental package as well.

File: modules/social_features/social_follow_taxonomy/modules/social_follow_landing_page/social_follow_landing_page.info.yml (hunk 4 above)
Comment (denysyukm) [reply modules/social_features/social_follow_taxonomy/modules/social_follow_landing_page/social_follow_landing_page.info.yml]: DONE

File: modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.info.yml (hunk 5 above)
Comment (denysyukm) [reply modules/social_features/social_follow_taxonomy/modules/social_follow_tag/social_follow_tag.info.yml]: DONE

**Comment by denysyukm** [comment]
@ronaldtebrake , I did continue working on the follow tag feature after Volodymyr and did some changes in module structure, functionality, etc. I based my work on your comment (commented on Jul 29). Could you review the changes, please. 

Let's prepare the PR for release!

**Comment by ronaldtebrake** [comment]
Hi guys, thanks for the update

1. I'm not entirely sure what the social_follow_taxonomy does without social_follow_tag. When I only enable sociall_follow_taxonomy I can already follow content tags it seems. Does this mean the follow_taxonomy is there for example for topic types and event types and such?
And the follow_tag module is there for when social_tagging is enabled?

Would be great to see if we can do something with this, can imagine it's not easy to understand when to enable what modules. Maybe even automatically enable social_follow_tag when social_tagging is enabled and social_follow_taxonomy. 

If we can't do that at least update the info.yml files to cleary reflect this.

2. Going to the search of a tag now indeed doesn't leak information nicely fixed, but it also shows no results for me, while i should see something. Also after cache clear and re-index, so it seems there is something not right with the indexing, which could be a local thing since I can't see it on eci_demo, but I'm also not sure what could cause that difference.

Screenshot 2020-12-30 at 13 54 33

3. FYI - I needed https://github.com/goalgorilla/open_social/pull/2080 to fix an issue with 
`You have requested a non-existent service "cache_context.config". Did you mean one of these: "cache_context.route`

4. After applying a sort or filter to the result set, the following bar disappears while im still looking at the other content, is that by design? Seems a bit weird since I am still in the context of the following tag.
https://www.loom.com/share/38cb4e865e25479e96b584f458fbb329

5. The button should probably behave the same as the dropbutton on adding members, it now looks a bit off. Also the caret doesn't change up / down.

Screenshot 2020-12-30 at 14 04 17

6. The size of the bar on the search should probably span the entire width of the hero / secondary tabs. 
Screenshot 2020-12-30 at 14 09 06

7. After adding a landing page, than enabling social_follow_landing_page and trying to add a tag to the section I get a
```
a:6:{s:5:"%type";s:24:"InvalidArgumentException";s:8:"@message";s:27:"Field field_tag is unknown.";s:9:"%function";s:58:"Drupal\Core\Entity\ContentEntityBase->getTranslatedField()";s:5:"%file";s:63:"/var/www/html/core/lib/Drupal/Core/Entity/ContentEntityBase.php";s:5:"%line";i:587;s:17:"@backtrace_string";s:6149:"#0 /var/www/html/core/lib/Drupal/Core/Entity/ContentEntityBase.php(568): Drupal\Core\Entity\ContentEntityBase->getTranslatedField('field_tag', 'x-default')
#1 /var/www/html/profiles/contrib/social/modules/social_features/social_follow_taxonomy/modules/social_follow_landing_page/social_follow_landing_page.module(169):```

https://www.loom.com/share/5cacd13a4ccf4378805bb89eb65a7e7f

I have also reproduced it here:
https://x-bitbucket-pr-105-sedgn4a-vncdokyquw6mc.eu.platform.sh/node/462/edit

8. Can we explain a bit better to our users what to do with landing pages and tagging content? Add tag leads you to this:
![Screenshot 2020-12-30 at 14 20 51](https://user-images.githubusercontent.com/16667281/103353990-a2bb1f00-4aaa-11eb-891b-8714ba5a9491.png)

Maybe we can explain what happens, how many items are shown (is it all items related to all the content, sorted by what?)

9. Have we got any performance metrics on this? I'm a bit worried when we get a lot of tags. All the preprocessing is pretty heavy.

10. After following a tag the block doesn't get updated so the star isn't shown
![Screenshot 2020-12-30 at 14 47 50](https://user-images.githubusercontent.com/16667281/103355225-07c44400-4aae-11eb-9250-ecd347257987.png)

**Comment by denysyukm** [comment]
@ronaldtebrake , thanks for a feedback. Let me answer on some points and ask some questions.

1. We divided social_follow_tag module from social_follow_taxonomy to have the possibility extend following taxonomy functionality on other taxonomies (not only tagging). 

Also, I already implemented automatically enabling social_follow_tag module when social_tagging is enabled and social_follow_taxonomy (see social_follow_taxonomy/social_follow_taxonomy.install).

2. Can you reproduce the search problem on the env https://x-bitbucket-pr-105-sedgn4a-vncdokyquw6mc.eu.platform.sh , please. Looks like search looks fine there (see: https://x-bitbucket-pr-105-sedgn4a-vncdokyquw6mc.eu.platform.sh/search/content?multistakeholder%5B%5D=24 )

3. Agree

4. Good catch. Needs to check what's going wrong there. Thank you!

5. Not sure it's related follow taxonomy feature and appeared through social_follow_taxonomy module code. Needs check.

6. Agree. Needs fix that. Thank you!

7. Will take a look on the bug. Thank you!

8. Agree with you. Whom can I ask to help me with texting?

9. I should ask the team about performance matrix. If I'm not mistaken someone else worked on that.

10. Good catch. Will take a look on that. Thank you!

**Comment by ronaldtebrake** [comment]
Hi @denysyukm thank you so much for your answers!

1) Sounds good! :) Think that indeed makes sense, just couldn't catch it at first sight when going through the info.yml files

2) I can't reproduce it there all seems to work well, however on distro nightly I can reproduce something else weird which might be related. 

I basically now see all results, not the filtered one. (login as benflorez who is a SM [here](https://x-bitbucket-pr-39-esftbaq-72phd64vv6oao.eu.platform.sh/search/content?tag1%5B%5D=8))

Screenshot 2021-01-05 at 09 11 51

8) I think we can ask xinyu to take a look, it's not a blocker for me can be added as a follow-up as well

9) Thank you, yeah it's more out of curiosity since I can imagine it's quite heavy with large content / taxonomy datasets

**Comment by ribel** [comment]
@denysyukm As we discussed with @ronaldtebrake, I have merged this PR.
Please create another one for the fixes mentioned above.
Also, we still need to prepare a release note.

**Pull request merged**